flink-python/pyflink/datastream/data_stream.py

```diff
@@ -895,6 +895,20 @@ def get_side_output(self, output_tag: OutputTag) -> 'DataStream':
         """
         return DataStream(self._j_data_stream.getSideOutput(output_tag.get_java_output_tag()))
 
+    def cache(self) -> 'DataStream':
+        """
+        Cache the intermediate result of the transformation. Only support bounded streams and
+        currently only block mode is supported. The cache is generated lazily at the first time the
+        intermediate result is computed. The cache will be clear when the StreamExecutionEnvironment
+        close.
+
+        :return: The cached DataStream that can use in later job to reuse the cached intermediate
+                 result.
+
+        .. versionadded:: 1.16.0
+        """
+        return DataStream(self._j_data_stream.cache())
```

Comment:
I guess we may need to return CachedDataStream which has some specific API, e.g. invalidate.

Reply:
On top of that, we also introduce `SideOutputDataStream` which supports the cache method. I think we may need to introduce `SideOutputDataStream` in PyFlink and let get_side_output return `SideOutputDataStream`.

Reply:
@Sxnan I think we can reuse the `cache` method of `DataStream`, no need to introduce a new class `SideOutputDataStream`